### ISSUE TITLE
Add available condition to control plane provider contract

### ISSUE THREAD

**Issue by ncdc** [kubernetes-sigs/cluster-api#3779]
**User Story**

As a developer, I would like to know when the control plane is first reachable for a new cluster, so that I can make decisions about MachineHealthCheck.spec.nodeStartupTimeout for joining control plane machines.

**Detailed Description**

When trying to determine if a joining control plane machine has exceeded a MachineHealthCheck's nodeStartupTimeout, we need to delay checking for this condition until after the control plane is first reachable. In other words, a joining control plane machine must not be marked unhealthy by the MachineHealthCheck if the apiserver is unreachable / it's still be initialized. After the apiserver is reachable, the nodeStartupTimeout should most likely be measured from when the machine's infrastructure was marked ready.

This proposal is to add a new required status condition to all control plane providers, `Available`, that indicates the first control plane node has completed bootstrapping and the apiserver is reachable. This is currently present in the KubeadmControlPlane provider, but it's not part of the control plane provider contract.

See also #3026

/kind feature
/area control-plane
/kind api-change

**Comment by ncdc** [comment]
Thinking a bit more, I wonder if what I really want is a Cluster condition for ControlPlaneAvailable, and the ClusterReconciler can set this when it's able to talk to the workload cluster apiserver? I don't know this necessarily or strictly needs to be in the control plane provider?

**Comment by detiber** [comment]
+1 to making this a condition on the Cluster

**Comment by vincepri** [comment]
That's what KubeadmControlPlane actually does today (more or less), we could write a fallback to do this generically by pinging the control plane?

**Comment by vincepri** [comment]
Initialized was a bit different than availability when we originally discussed it. If our goal is to provide something generic for saying "the control plane is available", consider something like:

- Have an `ControlPlaneAvailable` condition
- Have code that checks if the control plane reference conditions has `Available` set to `True`
- If the control plane reference isn't available, fallback to a simple connection check

What do you think?

**Comment by ncdc** [comment]
Hmm what does "available" mean? 😄 Let me come back later with some more detailed thoughts...

**Comment by vincepri** [comment]
`Initialized` could probably go away, it's currently set when a control plane is initialized, but doesn't necessarily mean that the control plane is ready to be taking requests or that's reachable.

`Available` was a way to say the control plane is ready to take requests and it's operational (needs more definition), which comes a bit after initialization step

**Comment by ncdc** [comment]
Going to move some of this discussion over to #3026 and will circle back here later

**Comment by detiber** [comment]
> `Initialized` could probably go away, it's currently set when a control plane is initialized, but doesn't necessarily mean that the control plane is ready to be taking requests or that's reachable.

Yes, it was initially intended mostly as a stop gap for the purposes of allowing migration and continuation of support for v1alpha2 Machine-based control planes to v1alpha3 and KubeadmControlPlane.

> `Available` was a way to say the control plane is ready to take requests and it's operational (needs more definition), which comes a bit after initialization step

My initial thoughts around this were to provide a signal that the ControlPlane (API Server) should be available for requests, so that it could block/unblock operations that require the control plane being available (such as creating worker Machines) in a way that is more accurate than `Initialized`, but not as strict as `Ready`. I would expect `Available` to be a subset of `Ready`, and should be able to switch between states (unlike `Initialized`).

I think `Available` should likely comprise the following:
- We have a control plane endpoint (supplied by user or a control plane provider)
- The apiserver is available through the endpoint (/healthz reports ok)

Eventually it might also be good to determine if the apiserver is in a state where resources can be created/modified, in which case it might be good to bubble up a signal from a control plane provider (if used). For KCP, this could simply bubble up that etcd has quorum. This could fall back to either assuming things are good or creating/deleting an arbitrary resource in the case that a control plane provider is not used.

**Comment by fabriziopandini** [comment]
The original ask

> As a developer, I would like to know when the control plane is first reachable for a new cluster, so that I can make decisions about MachineHealthCheck.spec.nodeStartupTimeout for joining control plane machines.

Has been already addressed in https://github.com/kubernetes-sigs/cluster-api/blob/c19ca28f5a43117d125237f4c44335759923d5ff/internal/controllers/machinehealthcheck/machinehealthcheck_targets.go#L149-L158 by taking the latest between the 3 times:
- machineCreationTime
- Cluster.InfrastructureReady transition to true (comes from infraCluster.status.ready)
- Cluster.ControlPlaneInitialized transition to true (comes from ControlPlane.status.initialized, falls back to the first time a CP machine reports nodeRef assigned when using standalone CP machines)

WRT to Initialized vs Available:
- As of today Cluster.ControlPlaneInitialized is used everywhere.
- Cluster.ControlPlaneInitialized derives from KCP.status.initialized, and it is up to the control plane provider to determine when set it to true (KCP currently assume initialized when there is a kubeadm-config map in the workload cluster).
- Available condition exists only in KCP, and it has the same meaning of KCP.status.initialized

Considered that IMO the most practical way forwards is 
- To close this issue because we don't need to change the contract, controlplane.status.initialized is already documented
- To open a new issue to drop KCP.Available, or to replace it with KCP.Initialized to get rid of the less used noun

@sbueringer @chrischdi opinions?

**Comment by sbueringer** [comment]
> To close this issue because we don't need to change the contract, ControlPlane.status.initialized is already documented

Sounds good

> To open a new issue to drop KCP.Available, or to replace it with KCP.Initialized to get rid of the less used noun

Sounds good to me, with the next apiVersion obviously.

**Comment by fabriziopandini** [comment]
/close

> The only thing that is not ideal is that this basically goes slightly backwards on our transition from status fields to conditions (the transition doesn't actually happen at the moment though).

Agreed, added a comment on https://github.com/kubernetes-sigs/cluster-api/issues/10532:
> My preference goes to introduce a KCP.Initialized condition (that mirrors ControlPlane.status.initialized)
